### ISSUE TITLE
Deglobalise rclone config

### ISSUE THREAD

**Issue by DigitalMarc** [rclone/rclone#4685]
Hi, 
I'm using rclone as a library but so far I had to prevent concurrent access (rclone config updates).
I was wondering if there is a plan to make rclone usable with various configs in parallel.

Thanks!

**Comment by ncw** [comment]
# Deglobalise rclone config

At the moment having global config is limiting rclone's ability to run different processes concurrently.

This is not noticeable in the command line, but when using rclone via the API or as a library it is very noticeable.

## Globals which need fixing

Rclone global config is registered by `rc.AddOption` and these can be seen using the rclone API with

    rclone rc --loopback options/get

The top level general config blocks are these - these are used for all rclone commands more or less.

- filter - filters in use for listing
- main - configures the general non-backend flags
- log - configures --log-file/format/syslog
- rc - configures remote control server (and `rclone rcd`)

These config blocks are used only by the relevant commands

- dlna - configures `rclone serve dlna`
- ftp - configures `rclone serve ftp`
- http - configures `rclone serve http`
- mount - configures `rclone mount`
- sftp - configures `rclone serve sftp`
- vfs - configures vfs flags for rclone mount/serve

## High level plan

Create a system to attach the current configuration to a `context.Context`.

Wherever global config is used, fetch it from the context, so instead of reading config like this

    fs.Config.DryRun

Read it with

    fs.GetConfig(ctx).DryRun

Where `fs.GetConfig` will read the ctx looking for a key for the main config. If one isn't found then it will use the (now private) default config.

This will need functions to create the new config contexts - maybe with a merge non-default values into the new config. This will enable use of rclone as a library with non-global config. So something like

    myConfig := fs.NewConfig()
    myConfig.DryRun = true // set custom config here
    newCtx := fs.SetConfig(ctx)

This will need a similar Set/Get function for each config block, however only the global config is widely spread througout the rclone code so there may be a neater way of doing it. This needs a bit of thought.

As part of this plan the now global config variables should be made package private so they can't be read from different packages.

## Potential difficulties

### Context everwhere

Rclone was started (at the time of go 1.0!) well before the `context.Context` system was invented. There has been some effort threading context through rclone, but there are likely more places that it needs to go. So for this change to be successful it will be necessary to thread context through more of rclone.

This is by and large a mechanical process and with the right editor support it doesn't take too long.

It will generate a lot of changes though!

See: https://github.com/rclone/rclone/issues/3257 for background.

### Config everywhere

`fs.Config` is used in 91 files in rclone so doing the change to use a context driven configuration will generate a lot of changes. Again these are mostly mechanical in nature and probably can be scripted.

`filter.Active` is used in 13 files.

## Extension work

### Use command line flag strings to create the context

So instead of the above, you'd write something like

    newCtx := config.FromFlags(ctx, `--dry-run --bwlimit 10M --exclude "*.jpg"`)

This would enable use from the API which would make the API much more usable

These flags could then go in the backend config say as a `flags` section for each backend.

**Comment by ncw** [comment]
I've pushed the work in progress to the [deglobalise](https://github.com/rclone/rclone/tree/deglobalise) branch.

For creating your own config use `fs.AddConfig`. This doesn't affect the filter config yet - that is next on the list.

```
func AddConfig(ctx context.Context) (context.Context, *ConfigInfo)
    AddConfig returns a mutable config structure based on a shallow copy of that
    found in ctx and returns a new context with that added to it.
```

There may be bugs! Testing appreciated :-)

**Comment by ncw** [comment]
I've merged the deglobalise branch to master now.

This contains

fs.GetConfig - to read the current config

https://github.com/rclone/rclone/blob/c22d04aa309c606cec7a7ac248df2fb4ee5aae15/fs/config.go#L171-L172

fs.SetConfig - to create a mutable config

https://github.com/rclone/rclone/blob/c22d04aa309c606cec7a7ac248df2fb4ee5aae15/fs/config.go#L183-L186

filter.GetConfig - to read the current filter config

https://github.com/rclone/rclone/blob/c22d04aa309c606cec7a7ac248df2fb4ee5aae15/fs/filter/filter.go#L602-L603

filter.SetConfig - to create a mutable filter config

https://github.com/rclone/rclone/blob/c22d04aa309c606cec7a7ac248df2fb4ee5aae15/fs/filter/filter.go#L614-L617

filter.ReplaceConfig - to repace the filter config with that passed in - this is probably the preferred way of setting filter config

https://github.com/rclone/rclone/blob/c22d04aa309c606cec7a7ac248df2fb4ee5aae15/fs/filter/filter.go#L625-L630

**Comment by joy4eg** [comment]
Hello, it's also would be nice to have a logger attached to the context (_not_ the global one).
And as the second part: triggers/callbacks to know what's going on with the particular transfer :)

**Comment by ncw** [comment]
> Hello, it's also would be nice to have a logger attached to the context (_not_ the global one).

Yes that did occur to me to, I didn't do it here in the interest of not breaking all the code at once!

> And as the second part: triggers/callbacks to know what's going on with the particular transfer :)

What sort of things do you want to know about?

**Comment by joy4eg** [comment]
>What sort of things do you want to know about?

I want to know when some file was transferred and call external API with that filename/details.

**Comment by macklin-10x** [comment]
@ncw this work looks great!  We're using rclone as a library and this will make it easier to potentially run rclone operations in parallel in the future.  Could you provide a quick summary of what would need to be tweaked to migrate to this new deglobalised config and filters?  For instance, right now we set `filter.Active` - is this passed in the context now?

**Comment by ncw** [comment]
@macklin-10x wrote:
> this work looks great! We're using rclone as a library and this will make it easier to potentially run rclone operations in parallel in the future. Could you provide a quick summary of what would need to be tweaked to migrate to this new deglobalised config and filters? For instance, right now we set `filter.Active` - is this passed in the context now?

What you do is call `filter.ReplaceConfig` with the filter that you made and then pass the `ctx` into all the calls. This is generally because when using filters you want to replace all of them.

https://github.com/rclone/rclone/blob/1406b6c3c9e21f59cba027101bfb2e014a21736e/fs/filter/filter.go#L642-L644

Whereas with the global config, what you want to do is call `fs.AddConfig`

https://github.com/rclone/rclone/blob/1406b6c3c9e21f59cba027101bfb2e014a21736e/fs/config.go#L205-L208

And adjust the config values in the returned `*ConfigInfo` to your satisfaction.

You can chain these calls so pass the `ctx` from the first into the second.

**Comment by ncw** [comment]
This should be all included in v1.55 now, so I'm going to close this issue.

**Comment by ncw** [comment]
> Thanks for your guidance, I've mostly managed to rebase our fork onto 1.55.0. I have two questions that have come up:
> 
> 1. The Fs constructors now take a context and often seem to store that context internally.  The fundamental sync operations (like `syncCopyMove`) also take a context.  Does the Fs pluck the current filter from the context with which it was constructed, or from the context passed into `syncCopyMove`?

The `Fs` only uses the filter for things like `--low-level-retries`. Filters are used from the one passed into `syncCopyMove`. They can of course be the same one if you want.

> 2. It looks like there is now an explicit call needed to load the config file, as my test suite is failing with many errors like this: `internal error: no config file system found. Did you call  ?`
>    Right now our application fakes a config file by assigning to the `ConfigFileGet` and `ConfigFileSet` function pointers.  Is there a call I can make to satisfy the above error without actually loading anything from the file system?

You can do several things...

The easiest thing for you would be to call `SetConfigPath` with an empty string indicating no config file, then call `configfile.LoadConfig(ctx)` as well as overriding `ConfigFileGet` and `ConfigFileSet` as you do at the moment

https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/fs/config/config.go#L317

Or you can implement the richer `Storage` interface

https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/fs/config/config.go#L68

And then call `SetStorage` to set it.

https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/fs/config/config.go#L333

You don't need to implement all the methods - you can just copy the skeleton here

https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/fs/config/default_storage.go#L3

And implement `GetValue` and `SetValue` should be enough unless you are going to call `rclone config`

**Comment by macklin-10x** [comment]
Awesome, thank you!  I ended up going the "implement `Storage`" route and that is working beautifully.  I ended up implementing the full Storage interface because it was straightforward to do in-memory; I could push it upstream as a PR that may be useful to future users embedding rclone as a library, though the implementation was so straightforward it might not be worth it.

**Comment by ncw** [comment]
Great!

If you'd like to send the in memory Storage that would be great. Quite possibly that should be the default and it is up to users to switch it out to something else.
